runtime/onert/core/src/backend/controlflow/kernel/WhileLayer.cc

```diff
@@ -29,57 +29,149 @@ namespace controlflow
   // // Run cond subg
-  // Copy cond subg inputs -> _dst_tensors
+  // If there is no loop copy "_input_tensors" -> "_dst_tensors", else copy "cond subg inputs" ->
+  // "_dst_tensors"
   auto cond_exec = dynamic_cast<exec::ExecutorBase *>(_executor_map->at(_cond_subg_index).get());
   auto body_exec = dynamic_cast<exec::ExecutorBase *>(_executor_map->at(_body_subg_index).get());
   if ((cond_exec == nullptr) || (body_exec == nullptr))
   {
     throw std::runtime_error{"While: Invalid condition or body"};
   }
 
-  const auto &cond_input_tensors = cond_exec->getInputTensors();
+  const auto &cond_graph = cond_exec->graph();
   const auto &cond_inputs_dyn_alloc = cond_exec->getInputsDynamicAllocInfo();
+  const auto &body_graph = body_exec->graph();
+  const auto &body_inputs_dyn_alloc = body_exec->getInputsDynamicAllocInfo();
+
+  std::vector<std::shared_ptr<backend::ITensor>> input_tensors;
+  std::vector<std::shared_ptr<backend::ITensor>> cond_input_tensors;
+  std::vector<std::shared_ptr<backend::ITensor>> body_input_tensors;
+  std::vector<std::shared_ptr<backend::ITensor>> body_output_tensors;
+  std::vector<std::shared_ptr<backend::ITensor>> output_tensors;
+
+  // Add only used tensors in cond subgraph
+  assert(_input_tensors.size() == cond_exec->getInputTensors().size());
+  for (uint32_t i = 0; i < cond_graph.getInputs().size(); ++i)
+  {
+    const auto &cond_input = cond_graph.operands().at(cond_graph.getInputs().at(i));
+    if (cond_input.getUses().size() > 0)
+    {
+      input_tensors.emplace_back(_input_tensors.at(i));
+      cond_input_tensors.emplace_back(cond_exec->getInputTensors().at(i));
+    }
+  }
   const auto permute_op_input_to_cond_input =
-      std::make_shared<PermuteLayer>(_input_tensors, cond_input_tensors, cond_inputs_dyn_alloc);
+      std::make_shared<PermuteLayer>(input_tensors, cond_input_tensors, cond_inputs_dyn_alloc);
 
-  const auto &body_input_tensors = body_exec->getInputTensors();
-  const auto &body_inputs_dyn_alloc = body_exec->getInputsDynamicAllocInfo();
-  const auto permute_cond_input_to_body_input =
-      std::make_shared<PermuteLayer>(cond_input_tensors, body_input_tensors, body_inputs_dyn_alloc);
+  // Add only used tensors among outputs of while operation
+  assert(_input_tensors.size() == _output_tensors.size());
+  assert(_output_indices.size() == _output_tensors.size());
+  input_tensors.clear();
+  output_tensors.clear();
+  for (size_t i = 0; i < _output_tensors.size(); ++i)
+  {
+    const auto &output_index = _output_indices.at(i);
+    const auto &output = _graph.operands().at(output_index);
+    if (output.getUses().size() > 0 || _graph.getOutputs().contains(output_index))
+    {
+      input_tensors.emplace_back(_input_tensors.at(i));
+      output_tensors.emplace_back(_output_tensors.at(i));
+    }
+  }
+  const auto permute_op_input_to_op_output =
+      std::make_shared<PermuteLayer>(input_tensors, output_tensors, _outputs_dyn_alloc_info);
```

Comment:
I am not sure if `permute_op_input_to_op_output` is really used. I am tracing this variable but it does not seem to be used.

Reply:
The variable is assigned at `auto permute_to_outputs_fn = permute_op_input_to_op_output;` and used at `permute_to_outputs_fn->run();`

Reply:
/to others
I asked the reason of of `permute_op_input_to_op_output` and @ragmani mentioned that `permute_op_input_to_op_output` is needed when cond is false and input needs to be copied into output.